lsコマンドの作成

sht3ng opened the pull request:
メソッドの組み合わせによってlsコマンドの機能を実装しました。
必須要件は対応済みです。
歓迎要件のコマンドライン引数は対応済みです。
歓迎要件の日本語対応は未対応です。

sht3ng committed:
lsコマンドの作成

メソッドの組み合わせによってlsコマンドの機能を実装しました。
必須要件は対応済みです。
歓迎要件のコマンドライン引数は対応済みです。
歓迎要件の日本語対応は未対応です。

yuuu commented on 05.ls/ls.rb:
```diff
@@ -0,0 +1,81 @@
+
+def calculate_row_count(files:)
+  column_count = calculate_column_count(files: files)
+  file_count   = calculate_file_count(files: files)
```
ここはメソッドに切り出さなくても、 `files.size` をそのまま活用した方がシンプルですね。

yuuu replied:
[こちら](https://github.com/sht3ng/ruby-practices/pull/4/files#diff-8d377e81683a20ae5ea549de93212074e856622b2543e434269d42607f38701bR44) や [こちら](https://github.com/sht3ng/ruby-practices/pull/4/files#diff-8d377e81683a20ae5ea549de93212074e856622b2543e434269d42607f38701bR55) も同様ですね。

sht3ng replied:
コメントありがとうございます😄
もしファイルのサイズを求める方法が変更になったらメソッドに切り出したほうが楽に変更できるかなと思ってこのようにしたのですが，よく考えたらファイルのサイズを求める方法に変更が求められることってあまりなさそうですね。笑
こちらと他の箇所も合わせて変更しておきたいと思います！😄

yuuu commented on 05.ls/ls.rb:
```diff
@@ -0,0 +1,81 @@
+def calculate_row_count(files:)
+  column_count = calculate_column_count(files: files)
+  file_count   = calculate_file_count(files: files)
+  if (file_count % column_count).zero?
```
`Float#ceil` を使うともっとシンプルに書けると思います 👍 (修正するかどうかはおまかせします)

sht3ng replied:
`Float#ceil`使うとシンプルですね！ありがとうございます！😄

sht3ng replied:
`Float#ceil`使うとシンプルですね！修正します！ありがとうございます！😄

yuuu commented on 05.ls/ls.rb:
```diff
@@ -0,0 +1,81 @@
+end
+
+def calculate_row_count(files:)
+  column_count = calculate_column_count(files: files)
```
前の処理で既にカラム数は計算済だと思うので、引数で受け取るようにするのはいかがでしょうか？

sht3ng replied:
もし他の人がこの関数を使うことになったらいろいろ引数があると面倒くさいかなと思ったことと，メソッド内で行数や列数を計算したら変数のスコープをできるだけ狭めることができるかなと思ったことからこのようにしたのですが，実行ごとで行数や列数が変更されることはないですし，それぞれのメソッド内で何回も計算するよりもメソッドの外で一度だけ行数や列数を計算する方がシンプルでよりわかりやすいですね！😄
こちらも変更させていただきます！😄

yuuu commented on 05.ls/ls.rb:
```diff
@@ -0,0 +1,81 @@
+  if file_count > MAX_COLUMN_COUNT
+    MAX_COLUMN_COUNT
+  else
+    file_count
```
今の処理内容を見る限り、ファイル数が `1` or `2` の場合も `3` を返してしまって良いように見えました。そうなるとこのメソッドは `3` を返すだけなので、削除しても良いかもしれません。

sht3ng replied:
こちらに関してなのですが，ファイル数が`1` or `2`の場合でも`3`を返すようにすると，`printf_files`メソッド内の`printf`メソッド実行時に`too few arguments (ArgumentError)`が出てしまいます💦
これは`printf_files`メソッド内の`printf`でフォーマット文字列（`format_specifier`）とフォーマットされる引数（`formatted_files`）の数が合わないことで発生するものです。

`generate_format_specifier`メソッドでは，`%-#{max_width}s`に`column_count`を掛け算することで，`printf_files`で使用するフォーマット文字列を作成しているため，ファイル数が`1` or `2`の場合でも`3`を返すようにすると，フォーマット文字列が`%-#{max_width}s * 3`となります。

`format_files`メソッドでは`row_count`（ファイル数が`1`や`2`のときには`1`）で分割した（`[1, 2] => [[1], [2]]`）あとにtransposeをする（`[[1], [2]] => [[1, 2]]`）ため，ファイル数が`1`もしくは`2`の場合は要素数が`1`もしくは`2`の配列の配列を`formatted_files`として返すようになっています。

この`format_specifier`と`formatted_files`の要素数を合わせるためには，`column_count`でフォーマット文字列の指示子の要素数を計算する方がシンプルかなと思ったのですが，こちらはこのまま修正しないでも大丈夫でしょうか。

そもそも今回のこの実装でよかったのか自信がないのですが，もしよりよい実装方法などがありましたら教えて頂けるととても嬉しいです😄

sht3ng replied:
こちらに関しまして，修正を行う中で，`printf_files`メソッド内の`printf`に渡すフォーマット文字列の指定子の個数とフォーマットされる引数の配列数で二人三脚のようなことを行っていたことを発見することができました！
これを避けるために`printf_files`メソッドで使用するフォーマット文字列を`each`文の外で作成していた方法から`each`文の中で作成する方法に変更を行なった結果，行数を定数で定義するだけでよくなりシンプルなコードにすることができました！
ありがとうございます！☺️

yuuu commented on 05.ls/ls.rb:
```diff
@@ -0,0 +1,81 @@
+  file_count   = calculate_file_count(files: files)
+  sliced_files = files.each_slice(row_count).to_a
+  unless (file_count % column_count).zero?
+    (row_count - sliced_files.last.size).times do
```
`each_slice` した後の最後の要素(配列)のサイズが3になるまで `nil` を加える、という書き方にすると分岐も不要になりますし、 `file_count` を事前に取得する必要もなくなりそうですね。

sht3ng replied:
おおー！この方法かなりスッキリしますね！😄
修正します！教えてくださってありがとうございます！😄

yuuu commented on 05.ls/ls.rb:
```diff
@@ -0,0 +1,81 @@
+  ls(path: path)
+end
+
+def ls(path:)
```
全体的にメソッドの引数をキーワード引数としているのは何か意図があってのことでしょうか？

sht3ng replied:
実は，はじめの方の実装ではメソッドの外から行数や列数を渡す方法にしていて，もし他の人がこの関数を使うことになったら順番を気にしなくていいことや可読性があがるかなと思ってキーワード引数にしていたのですが，今は引数がfilesひとつになってますし，キーワード引数の意味がなくなってますね。😅笑

キーワード引数は普通の引数指定よりも順番を気にしなくていいところと，メソッド実行時に引数の意味がわかりやすいことろがいいところだと思うのですが，引数が複数あるときの使い分け方で，もしなにかベストプラクティスのようなものがあれば教えていただけると嬉しいです！😄

yuuu replied:
そうですね、感覚値ですがメソッドを呼び出すのに必要な引数は通常の引数とすることが多いですね。
反面、オプション的なニュアンスで渡す引数をキーワード引数とすることが多いと思っています 👍

sht3ng replied:
教えてくださってありがとうございます！☺️
こういう感覚的なことって学びにくいものだと思いますので，教えてくださってとても助かります！☺️

sht3ng committed:
メソッドに切り出していたファイルのサイズを求める処理をfiles.sizeで求めるように変更しました。

sht3ng committed:
行数を求める処理をFloat#ceilを用いてシンプルにしました。

sht3ng committed:
メソッドの引数指定方法などの変更を行いました。

メソッドの引数指定をキーワード引数から普通の引数指定に変更しました。
メソッド内で使用する行数や列数などを外部で計算して引数で渡すように変更しました。

sht3ng committed:
行数をメソッドで計算する方法から定数で定義する方法に変更しました。

前回のコミット(0386ac1a476365b1d75d9d633c30279a86a78b20)でprintf_files内でprintfに渡すフォーマット文字列をeach文の中で生成するように変更したことで，行数をメソッドで計算する必要がなくなったため，定数で最大行数を定義する方法に変更しました。
また，これに伴って定数名をMAX_COLUMN_COUNTからCOLUMN_COUNTに変更しました。

sht3ng committed:
format_filesの処理方法を変更しました。

each_sliceした後の最後の要素のサイズが行数になるまでnilを加えるという方法に変更しました。

sht3ng committed:
テストを追加しました。

ファイルが0~6まであるときのテストと，正しくないパスを指定した時のテストを行っています。

sht3ng committed:
空のディレクトリを指定した時のエラーを修正しました。

ディレクトリが空のときにエラーが発生していたので，filesを取得したときに空だったら，lsメソッドから抜けるように処理を追加しました。

sht3ng committed:
read_files内の処理を変更しました。

エラーメッセージをOS標準のlsコマンドのエラーメッセージに近いものに修正しました。
ガード節を使用して，分岐を無くしました。
シンボリックリンクを指定した時のエラーを修正しました。
